src/chain/chain-training.cc

```diff
@@ -103,13 +104,57 @@ void ComputeChainObjfAndDeriv(const ChainTrainingOptions &opts,
     *l2_term = 0.0;
   } else {
     // compute the l2 penalty term and its derivative
-    BaseFloat scale = supervision.weight * opts.l2_regularize;
-    *l2_term = -0.5 * scale * TraceMatMat(nnet_output, nnet_output, kTrans);
-    if (nnet_output_deriv)
-      nnet_output_deriv->AddMat(-1.0 * scale, nnet_output);
+    BaseFloat scale_coeff = supervision.weight * opts.l2_regularize;
+    // If xent_output provided, l2 penalty is trying to regress the chain output
+    // to be a linear function of cross-entropy output.
+    // It minimizes -0.5 * l2_regularize * l2_norm(diag(scale) * x + offset - y)^2, 
+    // where x is cross-entropy output and y is chain output.
+    if (xent_output) {
+      //compute offset and scale
+      // The objecitve is to minimize L w.r.t scale_i, offset_i, 
+      // L = -0.5 * l2_regularize * 
+      //    \sum_{j=1}^{minibatch_size}(\sum_i (nnet_output_ji - target_ji)^2),
+      // where the target_ji = scale_i * xent_output_ji + offset_i. 
+      // scale_i = \sum_j (nnet_output_ji * xent_output_ji) / \sum_j(xent_output_ji^2)
```

Comment:
Dear Dan
Hi
No, I didn't optimize first w.r.t scale and then offset. I  jointly solved
them and I substitute offset as a function of scale in 1st equation and
solved scale.

On Thu, Jan 28, 2016 at 8:42 PM, Daniel Povey <notifications@github.com>
wrote:

> In src/chain/chain-training.cc
> <https://github.com/danpovey/kaldi/pull/4#discussion_r51216219>:
>
> > -    BaseFloat scale = supervision.weight * opts.l2_regularize;
> > -    *l2_term = -0.5 * scale * TraceMatMat(nnet_output, nnet_output, kTrans);
> > -    if (nnet_output_deriv)
> > -      nnet_output_deriv->AddMat(-1.0 * scale, nnet_output);
> > +    BaseFloat scale_coeff = supervision.weight * opts.l2_regularize;
> > +    // If xent_output provided, l2 penalty is trying to regress the chain output
> > +    // to be a linear function of cross-entropy output.
> > +    // It minimizes -0.5 * l2_regularize * l2_norm(diag(scale) * x + offset - y)^2,
> > +    // where x is cross-entropy output and y is chain output.
> > +    if (xent_output) {
> > +      //compute offset and scale
> > +      // The objecitve is to minimize L w.r.t scale_i, offset_i,
> > +      // L = -0.5 * l2_regularize *
> > +      //    \sum_{j=1}^{minibatch_size}(\sum_i (nnet_output_ji - target_ji)^2),
> > +      // where the target_ji = scale_i * xent_output_ji + offset_i.
> > +      // scale_i = \sum_j (nnet_output_ji * xent_output_ji) / \sum_j(xent_output_ji^2)
>
> i'm deleting your relevant jobs till this is fixed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/danpovey/kaldi/pull/4/files#r51216219>.
>